library/ssl_tls13_client.c

```diff
@@ -3726,6 +3729,13 @@ static int ssl_hrr_parse( mbedtls_ssl_context* ssl,
                         found = 1;
                         break;
                     }
+
+                    nr++;
+                    if( nr == MBEDTLS_SSL_MAX_KEY_SHARES )
```

Comment:
As I understand https://tools.ietf.org/html/rfc8446#section-4.2.8, the HRR does not contain a list of key shares, but only  single group ID. It's therefore not necessary to count the number of entries here.

Reply:
In this code segment I loop through the list of key shares sent in the original ClientHello. The TLS spec requires me to check whether the server indicated a key share in the HRR that was already included in the original ClientHello. If so, then something is wrong and the handshake will be terminated. 

You are, however, right that the check starting in line 3734 (regarding the MBEDTLS_SSL_MAX_KEY_SHARES) is not needed since the list of key shares is stored in ssl->handshake->key_shares_curve_list and is terminated with MBEDTLS_ECP_DP_NONE. 

Hence, I removed this part of the code.